storage/storage_transport.go

```diff
@@ -0,0 +1,294 @@
+		} else if len(storeInfo) == 2 && storeInfo[0] != "" && storeInfo[1] != "" {
+			// Two components: the driver type and the graph root.
+			if !filepath.IsAbs(storeInfo[1]) {
+				return nil, ErrPathNotAbsolute
+			}
+			store2, err := storage.GetStore(storage.StoreOptions{
+				GraphDriverName: storeInfo[0],
+				GraphRoot:       storeInfo[1],
+			})
+			if err != nil {
+				return nil, err
+			}
+			store = store2
+		} else {
+			// Anything else: store specified in a form we don't
+			// recognize.
+			return nil, ErrInvalidReference
+		}
+	}
+	return s.ParseStoreReference(store, reference)
+}
+
+func (s storageTransport) GetStoreImage(store storage.Store, ref types.ImageReference) (*storage.Image, error) {
+	dref := ref.DockerReference()
+	if dref == nil {
+		if sref, ok := ref.(*storageReference); ok {
+			if sref.id != "" {
+				if img, err := store.GetImage(sref.id); err == nil {
+					return img, nil
+				}
+			}
+		}
+		return nil, ErrInvalidReference
+	}
+	return store.GetImage(verboseName(dref))
+}
+
+func (s *storageTransport) GetImage(ref types.ImageReference) (*storage.Image, error) {
+	store, err := s.GetStore()
+	if err != nil {
+		return nil, err
+	}
+	return s.GetStoreImage(store, ref)
+}
+
+func (s storageTransport) ValidatePolicyConfigurationScope(scope string) error {
+	// Check if there's a store location prefix.
+	if scope[0] == '[' {
+		close := strings.IndexRune(scope, ']')
+		if close < 1 {
+			return ErrInvalidReference
+		}
+		storeSpec := scope[1:close]
+		scope = scope[close+1:]
+		storeInfo := strings.SplitN(storeSpec, "@", 2)
+		if len(storeInfo) == 1 && storeInfo[0] != "" {
+			// One component: the graph root.
+			if !filepath.IsAbs(storeInfo[0]) {
+				return ErrPathNotAbsolute
+			}
+		} else if len(storeInfo) == 2 && storeInfo[0] != "" && storeInfo[1] != "" {
+			// Two components: the driver type and the graph root.
+			if !filepath.IsAbs(storeInfo[1]) {
+				return ErrPathNotAbsolute
+			}
+		} else {
+			// Anything else: store specified in a form we don't
+			// recognize.
+			return ErrInvalidReference
+		}
+	}
+	// That might be all of it, and that's okay.
+	if scope == "" {
+		return nil
+	}
+	// But if there is anything left, it has to be a name, with or without
+	// an ID.
+	if scope[0] == '@' {
+		return ErrInvalidReference
+	}
+	scopeInfo := strings.SplitN(scope, "@", 2)
+	if len(scopeInfo) == 1 && scopeInfo[0] != "" {
+		id, name, err := reference.ParseIDOrReference(scopeInfo[0])
+		if err != nil {
+			return err
+		}
+		if id.Validate() == nil {
+			return ErrInvalidReference
+		}
+		if name == nil {
+			return ErrInvalidReference
+		}
+	} else if len(scopeInfo) == 2 && scopeInfo[0] != "" && scopeInfo[1] != "" {
+		id, name, err := reference.ParseIDOrReference(scopeInfo[0])
+		if err != nil {
+			return err
+		}
+		if id.Validate() == nil {
+			return ErrInvalidReference
+		}
+		if name == nil {
+			return ErrInvalidReference
+		}
+		id, name, err = reference.ParseIDOrReference(scopeInfo[1])
```

Comment:
The search path is (`requirementsForImageRef`): `PolicyConfigurationIdentity()` first, `PolicyConfigurationNamespaces()` in order, `""`. `ValidatePolicyConfigurationScope` should allow all used names on the path, otherwise they are rather pointless (it is OK if it accepts also other strings, it will “only” allow the user to write nonworking configuration without warning).

`PolicyConfigurationIdentity` can return an ID, so `ValidatePolicyConfigurationScope` should accept it.

But, yeah, as noted in the review summary, the way IDs don’t fit into reference namespacing at all made me give up on this for `docker-daemon:`.